### PR TITLE
Picopass: factory key support, app rename and move to NFC category, minor code cleanup

### PR DESCRIPTION
**Pull request by bettse**
# What's new

- Detect picopass factory default key
- Allow setting of iClass standard key
- App rename and move to NFC category, 
- Minor code cleanup

# Verification 

- Scan card that has picopass factory default debit key
- After success, select option for writing iClass standard key
- After success, re-read card and see that the regular read card result page is shown

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/plugins/picopass/scenes/picopass_scene_write_card_success.c [hunk 1]
```diff
@@ -16,6 +16,7 @@ void picopass_scene_write_card_success_widget_callback(
 void picopass_scene_write_card_success_on_enter(void* context) {
     Picopass* picopass = context;
     Widget* widget = picopass->widget;
+    FuriString* str = furi_string_alloc_set("Write Success!");
 
     DOLPHIN_DEED(DolphinDeedNfcReadSuccess);
```
Comment (bettse) [inline applications/plugins/picopass/scenes/picopass_scene_write_card_success.c:21]: for some reason this doesn't show up

File: applications/plugins/picopass/picopass_worker.c [hunk 2]
```diff
@@ -406,6 +438,84 @@ ReturnCode picopass_write_card(PicopassBlock* AA1) {
     return ERR_NONE;
 }
 
+ReturnCode picopass_write_block(PicopassPacs* pacs, uint8_t blockNo, uint8_t* newBlock) {
```
Comment (bettse) [inline applications/plugins/picopass/picopass_worker.c:441]: a lot of duplication with `write`, but not something I want to clean up right now

**Comment by bettse** [comment]
@pcunning / @nvx : thoughts?

**Comment by nvx** [comment]
Is the factory default key diversified? I thought the literal key was used raw?

As an unrealted issue, I noticed the description for `loclass_diversifyKey` is wrong here https://github.com/flipperdevices/flipperzero-firmware/blob/7980062b67a82dc3b3e9b74623b726251621dd7e/applications/plugins/picopass/lib/loclass/optimized_ikeys.c#L299-L305

That mentions it does the Elite KDF, but looks like it's really the standard KDF (which is used as the last step of the elite KDF mind you, but used alone is just the standard KDF). The elite KDF is here https://github.com/flipperdevices/flipperzero-firmware/blob/7980062b67a82dc3b3e9b74623b726251621dd7e/applications/plugins/picopass/lib/loclass/optimized_cipher.c#L283-L299

**Comment by bettse** [comment]
> Is the factory default key diversified? I thought the literal key was used raw?

I didn't think it was, but when testing the authentication it appeared to be 🤷‍♂️ .

**Comment by skotopes** [comment]
Writing iClass standard key gave me unreadable card. Could be that I did something wrong.

```870665 [D][PicopassWorker] csn 1dea8b05080012e0
870674 [D][PicopassWorker] config ffffffff7f1fffbc
870683 [D][PicopassWorker] aia ffffffffffffffff
870685 [E][PicopassWorker] Trying standard legacy key
870707 [D][PicopassWorker] rfalPicoPassPollerReadBlock 2 feffffffffffffff
870716 [D][PicopassWorker] rfalPicoPassPollerReadBlock 4 ffffffffffffffff
870727 [D][PicopassWorker] rfalPicoPassPollerReadBlock 5 ffffffffffffffff
870737 [D][PicopassWorker] rfalPicoPassPollerReadBlock 6 ffffffffffffffff
870746 [D][PicopassWorker] rfalPicoPassPollerReadBlock 7 ffffffffffffffff
870755 [D][PicopassWorker] rfalPicoPassPollerReadBlock 8 ffffffffffffffff
870764 [D][PicopassWorker] rfalPicoPassPollerReadBlock 9 ffffffffffffffff
870775 [D][PicopassWorker] rfalPicoPassPollerReadBlock 10 ffffffffffffffff
870785 [D][PicopassWorker] rfalPicoPassPollerReadBlock 11 ffffffffffffffff
870813 [E][PicopassWorker] rfalPicoPassPollerReadBlock error 4
870815 [E][PicopassWorker] picopass_read_card error 4
870821 [D][DolphinState] icounter 66, butthurt 0
873463 [D][DolphinState] icounter 66, butthurt 0
873502 [D][PicopassWorker] csn 1dea8b05080012e0
873514 [D][PicopassWorker] config ffffffff7f1fffbc
873526 [D][PicopassWorker] aia ffffffffffffffff
873528 [E][PicopassWorker] Trying standard legacy key
873550 [D][PicopassWorker] rfalPicoPassPollerReadBlock 2 feffffffffffffff
873559 [D][PicopassWorker] rfalPicoPassPollerReadBlock 4 ffffffffffffffff
873569 [D][PicopassWorker] rfalPicoPassPollerReadBlock 5 ffffffffffffffff
873579 [D][PicopassWorker] rfalPicoPassPollerReadBlock 6 ffffffffffffffff
873588 [D][PicopassWorker] rfalPicoPassPollerReadBlock 7 ffffffffffffffff
873597 [D][PicopassWorker] rfalPicoPassPollerReadBlock 8 ffffffffffffffff
873606 [D][PicopassWorker] rfalPicoPassPollerReadBlock 9 ffffffffffffffff
873615 [D][PicopassWorker] rfalPicoPassPollerReadBlock 10 ffffffffffffffff
873626 [D][PicopassWorker] rfalPicoPassPollerReadBlock 11 ffffffffffffffff
873636 [D][PicopassWorker] rfalPicoPassPollerReadBlock 12 ffffffffffffffff
873645 [D][PicopassWorker] rfalPicoPassPollerReadBlock 13 ffffffffffffffff
873654 [D][PicopassWorker] rfalPicoPassPollerReadBlock 14 ffffffffffffffff
873665 [D][PicopassWorker] rfalPicoPassPollerReadBlock 15 ffffffffffffffff
873674 [D][PicopassWorker] rfalPicoPassPollerReadBlock 16 ffffffffffffffff
873683 [D][PicopassWorker] rfalPicoPassPollerReadBlock 17 ffffffffffffffff
873692 [D][PicopassWorker] rfalPicoPassPollerReadBlock 18 ffffffffffffffff
873702 [D][PicopassWorker] rfalPicoPassPollerReadBlock 19 ffffffffffffffff
873712 [D][PicopassWorker] rfalPicoPassPollerReadBlock 20 ffffffffffffffff
873721 [D][PicopassWorker] rfalPicoPassPollerReadBlock 21 ffffffffffffffff
873730 [D][PicopassWorker] rfalPicoPassPollerReadBlock 22 ffffffffffffffff
873739 [D][PicopassWorker] rfalPicoPassPollerReadBlock 23 ffffffffffffffff
873750 [D][PicopassWorker] rfalPicoPassPollerReadBlock 24 ffffffffffffffff
873760 [D][PicopassWorker] rfalPicoPassPollerReadBlock 25 ffffffffffffffff
873769 [D][PicopassWorker] rfalPicoPassPollerReadBlock 26 ffffffffffffffff
873778 [D][PicopassWorker] rfalPicoPassPollerReadBlock 27 ffffffffffffffff
873788 [D][PicopassWorker] rfalPicoPassPollerReadBlock 28 ffffffffffffffff
873797 [D][PicopassWorker] rfalPicoPassPollerReadBlock 29 ffffffffffffffff
873806 [D][PicopassWorker] rfalPicoPassPollerReadBlock 30 ffffffffffffffff
873815 [D][PicopassWorker] rfalPicoPassPollerReadBlock 31 ffffffffffffffff
873817 [D][PicopassDevice] Unknown encryption
873819 [D][PicopassDevice] bitLength: 255
873823 [D][DolphinState] icounter 66, butthurt 0
```

**Comment by bettse** [comment]
>Writing iClass standard key gave me unreadable card.

The log doesn't seem to indicate that.  After trying the standard key it didn't fail and try the elite, instead this implies it succeeded and started to read the blocks.  If your card is anything like the unpersonalized (factory default) ones I hae, then it comes with all the blocks being all FF, so this output makes sense for all the non-key blocks.

If you get a "read error" in the UI after writing, that would be because it doesn't any PACS data (this could be improved on).  It should allow you to write the PACS of an existing card to it, which would then make it not show a read error.

**Comment by skotopes** [comment]
> > Writing iClass standard key gave me unreadable card.
> 
> The log doesn't seem to indicate that. After trying the standard key it didn't fail and try the elite, instead this implies it succeeded and started to read the blocks. If your card is anything like the unpersonalized (factory default) ones I hae, then it comes with all the blocks being all FF, so this output makes sense for all the non-key blocks.
> 
> If you get a "read error" in the UI after writing, that would be because it doesn't any PACS data (this could be improved on). It should allow you to write the PACS of an existing card to it, which would then make it not show a read error.

Yep, no PACS data was written. So if all non-key blocks are `FF` is it safe to say to user that card is empty instead of read error?

**Comment by bettse** [comment]
> Yep, no PACS data was written. So if all non-key blocks are FF is it safe to say to user that card is empty instead of read error?

100%, it is just not something I'd done in the past.  I can take that on in a subsequent PR

**Comment by bettse** [comment]
Given I had to fix the array indexing, I also added the empty detection

**Review by skotopes — APPROVED** [review]

**Pull request merged**